src/test/java/org/tarantool/jdbc/JdbcPreparedStatementIT.java

```diff
@@ -719,6 +719,161 @@ public void testSetBadCharacterStream() throws Exception {
+            assertEquals("two", resultSet.getString(1));
+            assertFalse(resultSet.next());
+        }
+    }
+
+    @Test
+    public void testLikeEscapeSyntax() throws Exception {
+        testHelper.executeSql("INSERT INTO test(id, val) VALUES (1, 'one%'), (2, 'two'), (3, 'three%'), (4, 'four')");
+
+        prep = conn.prepareStatement("SELECT val FROM test WHERE val LIKE '%|%' {escape ?}");
+        prep.setString(1, "|");
+
+        prep.execute();
+
+        try (ResultSet resultSet = prep.getResultSet()) {
+            assertTrue(resultSet.next());
+            assertEquals("one%", resultSet.getString(1));
+            assertTrue(resultSet.next());
+            assertEquals("three%", resultSet.getString(1));
+            assertFalse(resultSet.next());
+        }
+    }
+
+    @Test
+    public void testOuterJoinEscapeSyntax() throws Exception {
+        testHelper.executeSql("INSERT INTO test(id, val) VALUES (1, 'one')");
+
+        prep = conn.prepareStatement(
+            "SELECT {fn concat('t1-', t1.val)}, {fn concat('t2-', t2.val)} " +
+                "FROM {oj test t1 LEFT OUTER JOIN test t2 ON t1.id = ?}"
+        );
+        prep.setInt(1, 1);
+
+        prep.execute();
+
+        try (ResultSet resultSet = prep.getResultSet()) {
+            assertTrue(resultSet.next());
+            assertEquals("t1-one", resultSet.getString(1));
+            assertEquals("t2-one", resultSet.getString(2));
+            assertFalse(resultSet.next());
+        }
+    }
+
+    @Test
+    public void testSystemFunctionEscapeSyntax() throws Exception {
+        testHelper.executeSql("INSERT INTO test(id, val) VALUES (1, NULL)");
+
+        prep = conn.prepareStatement("SELECT {fn ifnull(val, ?)} FROM test WHERE id = 1");
+        prep.setString(1, "one-one");
+
+        prep.execute();
+
+        try (ResultSet resultSet = prep.getResultSet()) {
+            assertTrue(resultSet.next());
+            assertEquals("one-one", resultSet.getString(1));
+            assertFalse(resultSet.next());
+        }
+    }
+
+    @Test
+    public void testNumericFunctionEscapeSyntax() throws Exception {
+        testHelper.executeSql("INSERT INTO test(id, val) VALUES (1, NULL)");
+
+        prep = conn.prepareStatement("SELECT {fn abs(5 - ?)}, {fn round({fn pi()}, ?)}");
+        prep.setInt(1, 10);
+        prep.setInt(2, 0);
+
+        prep.execute();
+
+        try (ResultSet resultSet = prep.getResultSet()) {
+            assertTrue(resultSet.next());
+            assertEquals(5, resultSet.getInt(1));
+            assertEquals(3, resultSet.getInt(2));
+            assertFalse(resultSet.next());
+        }
+    }
+
+    @Test
+    public void testStringFunctionEscapeSyntax() throws Exception {
+        testHelper.executeSql("INSERT INTO test(id, val) VALUES (1, 'one'), (2, 'TWO'), (3, 'three'), (4, ' four ')");
+
+        prep = conn.prepareStatement(
+            "SELECT {fn char(?)}, {fn right(val, 2)}, {fn concat(?, val)} FROM test WHERE id = 3"
```

Comment:
well, we can't use `existing LENGTH() function` directly because of incompatible with the spec behavior;

Appendix C says:

> LENGTH(string) - Number of characters in a string, **excluding trailing blanks**

This is the reason we transform it into `LENGTH(TRIM(TRAILING FROM string))`. But `TRAILING FROM` is not supported in 2.1 too.

I moved a few v2.1-unsupported functions into separate tests.

https://github.com/tarantool/tarantool-java/blob/d1f72f28ea72b18d729774c4f5849ce3033dd67f/src/test/java/org/tarantool/jdbc/JdbcPreparedStatementIT.java#L887-L909

https://github.com/tarantool/tarantool-java/blob/d1f72f28ea72b18d729774c4f5849ce3033dd67f/src/test/java/org/tarantool/jdbc/JdbcStatementIT.java#L707-L726

Reply:
One more change is related to `RAND` function:

> RAND(integer) - Random **floating-point** for seed integer

Here, I set eyes on `floating-point` that it usually means a value within [0, 1) (according to other jdbc vendors too). It is also not the same as Tarantool `RANDOM()` function which returns integer value between [-2^63-1, 2^64], AFAIK. I was thinking of how we can overcome and normalize this value to the required range, say using smt like `abs(random())/max_int` but failed due to the formula extra complexity. So, Instead of using `RANDOM()` the driver uses its own PRNG (`new Random().nextDouble()`) to substitute invocation of the function by double literal directly.

Reply:
Re `LTRIM()` and `RTRIM()`: it was changed in https://github.com/tarantool/tarantool/issues/3879 since 2.2.0-198-gf95a34da9 (and for all 2.3 versions). tarantool-2.1 has syntax inherited from SQLite: `LTRIM()` and `RTRIM()`. I would make the replacement based on tarantool version.

Okay for `RAND()`.